[J-hta-n] iP

J-hta-n opened the pull request:
# Ekud 
> "Never gonna give you up" - Rick Astley ([source](https://youtu.be/dQw4w9WgXcQ?si=D0DqKudSbGmlBswu))

Ekud helps you to easily view all your priorities at a glance! It's:
- text based
- **simple** to learn
- ~~fast~~ **_REALLY_** fast to use

To use it, all you have to do is:
1. download the .jar file from either [here](https://github.com/J-hta-n/ip/releases/tag/A-Jar) or [here](https://youtu.be/dQw4w9WgXcQ?si=D0DqKudSbGmlBswu).
2. double-click it.
3. add or modify your tasks.
4. there you have it :joy:

Most of all, it's ***FREE*** to use!

Features:
- [x] Managing tasks, deadlines, and events
- [ ] GUI (coming soon)
- [ ] Reminders (coming soon)

You can try building this App to practise Java too! Here's the `main` method:
```java
public class Ekud {
  public static void main(String[] args) {
    Ekud.launch(Ekud.class, args);
  }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

J-hta-n committed:
Add initial skeletal version

J-hta-n committed:
Add echo functionality

J-hta-n committed:
Change echo to user text list functionality

J-hta-n committed:
Add mark as done functionality

J-hta-n committed:
Add support for different task types

J-hta-n committed:
Add test case (text UI testing)

J-hta-n committed:
Refactor design and add exception handling for invalid inputs

Made TaskList an object to handle task modifications and handling of illegal arguments for modifying the TaskList with the EkudIllegalArgException, while EkudInvalidCommandException is handled from the main chatbot program itself

J-hta-n committed:
Add delete functionality for task list

J-hta-n committed:
Update test case

J-hta-n committed:
Refactor code to handle all invalid inputs in TaskList

J-hta-n committed:
Touchup on documentation

Also tried using enum for commands, but not done since invalid commands are already handled with the generic EkudInvalidCommandException mesage

J-hta-n committed:
Minor refactor

J-hta-n committed:
Add enum for commands

J-hta-n committed:
Trivial modifications

J-hta-n committed:
Change constant naming

J-hta-n committed:
Add functionality of saving tasks in hard disk

Everytime a user quits the program, the tasks are saved and loaded up for the next session.

Exceptions with reading or writing of data is also handled to inform the user of the error.

J-hta-n committed:
Add formatting for date and times (in deadlines and events)

Program now requires users to input a valid dateTime format for parsing. A similar parsing process is used

when reading from the saved task list file.

J-hta-n committed:
Merge branch 'branch-Level-7'
Adds saving of tasks functionality to the program

J-hta-n committed:
Merge branch 'branch-Level-8'
Adds support for dateTime objects in the program

J-hta-n committed:
Refactor code by using more OOP

J-hta-n committed:
Merge branch 'branch-A-MoreOOP'

Uses more OOP classes, namely Parser, Ui and
Storage to improve code maintanability

J-hta-n committed:
Organise classes into packages

J-hta-n committed:
Merge branch 'branch-A-Packages'

Organises various classes into distinct packages, namely exceptions,
parser, storage, tasks and ui packages.

J-hta-n committed:
Merge branch 'add-gradle-support'

Adds Gradle wrapper files to the project.

J-hta-n committed:
Add initial JUnit Tests

J-hta-n committed:
Merge branch 'branch-A-JUnit'

Adds some initial basic JUnit testing for the Parser
and TaskList classes.

J-hta-n committed:
Create a JAR File

J-hta-n committed:
Merge branch 'branch-A-Jar'

Configures the App such that it can be packaged
as an executable JAR file.

J-hta-n committed:
Add more JavaDocs

J-hta-n committed:
Tweak code to comply with coding standard

Introduces linebreaks to ensure ~110 chars and strictly

< 120 chars per line. Adds spacing in between 'catch' and

'(Exception e)' too.

J-hta-n committed:
Add find tasks by keyword functionality

J-hta-n committed:
Merge branch 'branch-A-JavaDoc'

Adds more detailed JavaDocs for non-trivial private and public methods.

J-hta-n committed:
Merge branch 'branch-A-CodingStandard'

Refactors the code to adhere more strictly to the project's
coding standard, namely by adding line breaks and spacing
in between 'catch' and '(Exception e)' statements.

J-hta-n committed:
Merge branch 'branch-Level-9'

Resolved merge conflicts in Parser.java and TaskList.java
before merging.

J-hta-n committed:
Configure checkStyle to detect coding style violations

J-hta-n committed:
Merge branch 'branch-A-CheckStyle'

Now uses checkStyle as taken from
https://se-education.org/guides/tutorials/checkstyle.html
to detect coding style violations in real time in
IntelliJ IDE.

nikele2001 commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,110 @@
+        this.path = path;
+        File file = new File(path);
+        try {
+            if (!file.getParentFile().exists()) { // Create directory if it doesn't exist
```
Good job for considering whether directory where data file does not exists!

nikele2001 commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,207 @@
+     */
+    public void showTasks() {
+        if (this.tasks.isEmpty()) {
+            Ui.printMsg("Your to-do list is currently empty :o");
```
Is it a good idea to allow the TaskList class to use static methods of the Ui class? In my opinion, the Ui class should be the one calling these methods. However, this works so all's good!

nikele2001 commented on src/main/java/ui/Ui.java:
```diff
@@ -0,0 +1,46 @@
+     * Prints a message formatted in between 2 horizontal lines for the user.
+     * @param message Text to be printed.
+     */
+    public static void printMsg(String message) {
```
Instead of creating a generic printMsg method, i think it would be good to outline certain methods to allow the ui to print certain lines that are commonly printed (e.g. adding of tasks, marking, unmarking, deleting, etc.)

nikele2001 commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,207 @@
+     * Prints a confirmation message for the user after adding a task.
+     * @param task
+     */
+    private void printAddedTask(Task task) {
```
In my opinion, this method should exist in the Ui class.

nikele2001 commented on src/test/java/tasks/TaskListTest.java:
```diff
@@ -0,0 +1,30 @@
+
+public class TaskListTest {
+    @Test
+    public void deleteTask_emptyTaskList_throwsException() {
```
Great initiative to test whether user-defined exception class works! :)

nikele2001 reviewed (commented):
Overall, great job for putting together this program! There is no major issues in my opinion, but perhaps due to some nuances of certain OOP principles, there are some methods of printing lines that I feel should be in the Ui class instead of say the TaskList class. Otherwise, well done and keep up the good work! :)

Bearypop commented on src/main/java/Ekud.java:
```diff
@@ -0,0 +1,74 @@
+import exceptions.*;
```
Maybe import only the required exceptions?

J-hta-n committed:
Add JavaFX GUI files to ui package

Files taken from https://se-education.org/guides/tutorials/javaFxPart4.html.

Also shifted all pre-existing classes and packages into a new ekud package,

and removed the unused Duke.java file.

J-hta-n committed:
Refactor codebase to work with the JavaFX GUI

Removed the old command line UI. To work with
the GUI, all outputs printed via System.out.println()
in the old version now return a String response for
every user input instead. Ekud.java has been
modified in various places too.

J-hta-n committed:
Merge branch 'branch-Level-10'

Replaces the old command line interface with the
JavaFX GUI.

J-hta-n committed:
Use varargs

Uses varargs for the addEvent() function in TaskList.java for a start.

J-hta-n committed:
Merge branch 'branch-A-Varargs'

Introduces the use of varargs.

J-hta-n committed:
Add more varargs

J-hta-n committed:
Merge branch 'branch-A-Varargs'

Second merge for minor changes.

Bearypop commented on src/main/java/parser/Command.java:
```diff
@@ -0,0 +1,28 @@
+    ADDDEADLINE("deadline"),
+    ADDEVENT("event"),
+    DELETETASK("delete"),
+    FIND("find");
```
I like that you use caps here for the enums.

Bearypop commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,110 @@
+                    LocalDateTime toDateTime = parser.parseSavedDateTime(taskDetails[4]);
+                    taskList.addTask(new Event(taskDetails[2], fromDateTime, toDateTime));
+                }
+                if (isDone) taskList.markDoneOnStart(curTask);
```
The provided guide suggests writing it this way:
```suggestion
                if (isDone) {
                    taskList.markDoneOnStart(curTask);
                }
```

Bearypop reviewed (commented):
Overall, the code looks good. The codes follows almost all of the coding conventions. Nice!

J-hta-n committed:
Add assertions and some javadocs

Currently, there are no Java assertions in place to verify
assumptions about the chatbot's state at various points.

Given that the program is getting larger, it is becoming increasingly
difficult and time-consuming to identify bugs caused by invalid
assumptions or mistakes regarding the program's state.

Adding assertions helps to address this potential issue.

For a simple start, let's add some assertions to a few key methods
of the chatbot, namely its constructor and the main function for parsing
and executing commands. Future assertions shall be added progressively
according to evolving requirements.

J-hta-n committed:
Improve Code Quality

Codebase hasn't been refactored in a while, which may lead to unsafe
practises or readability issues, eventually leading to bugs.

Examining the code and refactoring it by adhering to code quality
guidelines helps to prevent technical debt from accumulating.

For this refactoring commit, let's
* ensure that all switch-case/if blocks have a default/else block
to handle invalid cases
* ensure guard clauses are used where applicable to prevent
unnecessary nesting and make the happy path prominent
* replace magic literals
* use enums where possible, namely the TaskType enum
* avoid complicated expressions

J-hta-n committed:
Merge pull request #1 from J-hta-n/branch-A-Assertions

Add assertions and some javadocs

J-hta-n committed:
Merge pull request #2 from J-hta-n/branch-A-CodeQuality

Improve Code Quality

J-hta-n committed:
Add Task Priority functionality

Currently, the lack of priorisation of tasks makes it
hard to begin assigning tasks to clear.

Let's assign a priority level to every task, mark it
clearly for the user, and allow the user to change the
priority level.

J-hta-n committed:
Improve JavaFX GUI

The current GUI looks disorganised with inadequate spacing.

Let's add some padding and margin to the DialogBox element,
and also clip the images to a circle.

J-hta-n committed:
Merge pull request #3 from J-hta-n/branch-C-Priority

Branch c priority

J-hta-n committed:
Add background color

User and Ekud dialog boxes now have their own background color
for an improveed GUI.

J-hta-n committed:
Merge pull request #5 from J-hta-n/branch-A-BetterGUI

Add background color

J-hta-n committed:
Minor refactor in MainWindow.java

J-hta-n committed:
Merge pull request #6 from J-hta-n/branch-A-BetterGUI

Minor refactor in MainWindow.java

J-hta-n committed:
Shorten long method in Parser.java

The parseAndExecute() method in Parser.java is currently too long,
which may impede readability.

Let's shorten the method by abstracting away details into helper
functions instead.

J-hta-n committed:
Merge pull request #7 from J-hta-n/branch-A-CodeQuality

Shorten long method in Parser.java

J-hta-n committed:
Add more error handling

Let's shift the error handling logic for index out
of bounds from taskList to parser to abide by separation
of concerns. More error handling such as end date time
being earlier than the start date time for events have
been added as well.

J-hta-n committed:
Merge pull request #9 from J-hta-n/branch-A-MoreErrorHandling

Add more error handling

J-hta-n committed:
Add User Guide and Screenshot

Adds a representative screenshot (Ui.png) and a basic
user guide for all importnat features of the chatbot.

J-hta-n committed:
Merge pull request #10 from J-hta-n/branch-A-UserGuide

Add User Guide and Screenshot

J-hta-n committed:
Update README.md

J-hta-n committed:
Update Screenshot

Modifies the current GUI by having user and chatbot
dialogs, and adds a new Ui.png screenshot

J-hta-n committed:
Merge pull request #11 from J-hta-n/branch-A-UserGuide

Update Screenshot

J-hta-n committed:
Improve code quality

It's been quite a few commits since the last code refactoring,
with readability issues regarding long methods and improper
javadocs.

Let's address the aforementioned issues by incorporating SLAP
to shorten two long methods, particularly the loadData() and
parseAndExecute() methods. Let's also refactor the JavaDocs
to ensure that their phrasing adheres to coding standards.

J-hta-n committed:
Merge pull request #12 from J-hta-n/branch-A-CodeQuality

Improve code quality

J-hta-n committed:
Add more JUnit test cases

Adds a few more automated JUnit tests for the parser
and taskList objects.

J-hta-n committed:
Merge pull request #13 from J-hta-n/branch-A-MoreTesting

Add more JUnit test cases

J-hta-n committed:
Refactor javadocs

Modifies some comments after proofreading the codebase
to ensure javadocs coding standards are adhered to as
much as possible in general.

J-hta-n committed:
Merge pull request #14 from J-hta-n/branch-A-CodeQuality

Refactor javadocs

J-hta-n committed:
Minor UI edit

Adjusts the alignment of the chatbot dialog to centre left.

J-hta-n committed:
Merge pull request #16 from J-hta-n/branch-A-BetterGUI

Minor UI edit

J-hta-n committed:
Update README.md

J-hta-n commented:
> Overall, great job for putting together this program! There is no major issues in my opinion, but perhaps due to some nuances of certain OOP principles, there are some methods of printing lines that I feel should be in the Ui class instead of say the TaskList class. Otherwise, well done and keep up the good work! :)

Thank you for all the feedback! :) Now I'm more mindful of where to allocate methods during OOP design 👍

J-hta-n commented:
> Overall, the code looks good. The codes follows almost all of the coding conventions. Nice!

Thank you all for the feedback once more! :) I'll be more careful when it comes to formatting in adhering to coding standards 👍

J-hta-n committed:
Add CI

Allows GitHub Actions to set up CI.

J-hta-n committed:
Merge pull request #17 from J-hta-n/branch-A-CI

Add CI

J-hta-n committed:
Remove IO tests from .yml

J-hta-n committed:
Update README.md

J-hta-n committed:
Update README.md

J-hta-n committed:
Update README.md

J-hta-n committed:
Update README.md

J-hta-n committed:
Update .yml

J-hta-n committed:
Update .yml
